python/rbac/endpoints_ui/role/views.py

```diff
@@ -0,0 +1,105 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Role object candidates view set"""
+
+from collections import defaultdict
+
+from rest_framework import serializers, mixins
+from rest_framework.response import Response
+from rest_framework.decorators import action
+
+from cm import models as cm_models
+from rbac import models
+from rbac.viewsets import GenericPermViewSet
+
+
+class RoleSerializer(serializers.Serializer):
+    id = serializers.IntegerField()
+    name = serializers.CharField()
+    parametrized_by_type = serializers.JSONField()
+    object_candidate_url = serializers.HyperlinkedIdentityField(
+        view_name='rbac-ui:role-object-candidate'
+    )
+
+
+class RoleViewSet(mixins.ListModelMixin, GenericPermViewSet):
+    queryset = models.Role.objects.filter(type=models.RoleTypes.role).all()
+    serializer_class = RoleSerializer
+
+    @action(methods=['get'], detail=True)
+    def object_candidate(self, request, **kwargs):
+        role = self.get_object()
+        clusters = []
+        providers = []
+        services = []
+        hosts = []
+
+        if models.ObjectType.cluster.value in role.parametrized_by_type:
+            for cluster in cm_models.Cluster.objects.all():
+                clusters.append(
+                    {
+                        'name': cluster.display_name,
+                        'type': cm_models.PrototypeEnum.Cluster.value,
+                        'id': cluster.id,
+                    }
+                )
+
+        if models.ObjectType.provider.value in role.parametrized_by_type:
+            for provider in cm_models.HostProvider.objects.all():
+                providers.append(
+                    {
+                        'name': provider.display_name,
+                        'type': cm_models.PrototypeEnum.Provider.value,
+                        'id': provider.id,
+                    }
+                )
+
+        if models.ObjectType.host.value in role.parametrized_by_type:
+            for host in cm_models.Host.objects.all():
+                hosts.append(
+                    {
+                        'name': host.display_name,
+                        'type': cm_models.PrototypeEnum.Host.value,
+                        'id': host.id,
+                    }
+                )
+
+        if (
+            models.ObjectType.service.value in role.parametrized_by_type
+            or models.ObjectType.component.value in role.parametrized_by_type
+        ):
+            _services = defaultdict(list)
+            for service in cm_models.ClusterObject.objects.all():
+                _services[service.display_name].append(
+                    {
+                        'name': service.cluster.display_name,
+                        'type': 'service',
+                        'id': service.id,
```

Comment:
```suggestion
                        'type': 'cluster',
                        'id': service.cluster.id,
```

Reply:
No, it is done as required - by selecting cluster from that list user selects instance of that cluster's service

Reply:
Yes, it dawned on me